benchmarks/MOABB/train.py

```diff
@@ -1,42 +1,181 @@
+    dataset.add_dynamic_item(
+        encode_label_func, takes=["label"], provides="encoded_label",
+    )
+
+    # 4) Change the dataset output keys to produce encoded_label instead of raw "label"
+    #    (You can keep "label" too if you want both.)
+    dataset.set_output_keys(["encoded_label", "subject", "session", "epoch"])
+
+    return dataset
+
+
+def prepare_splits(hparams, dataset):
+    """Create train/valid/test splits using new splitter system."""
+
+    # Create appropriate splitter
+    if hparams["data_iterator_name"] == "leave-one-session-out":
+        splitter = CrossSessionSplitter(dataset, leave_k_out=1)
+    elif hparams["data_iterator_name"] == "leave-one-subject-out":
+        splitter = CrossSubjectSplitter(dataset, leave_k_out=1)
+    else:
+        raise ValueError(f"Unknown split type: {hparams['data_iterator_name']}")
+
+    # Get specific split based on session index
+    split = list(splitter)[hparams["target_session_idx"]]
+    train_dataset = split["train"]
+    total_len = len(train_dataset)
+    val_ratio = hparams["valid_ratio"]
+    train_len = int(total_len * (1 - val_ratio))
+    val_len = total_len - train_len
+
+    generator = torch.Generator().manual_seed(hparams["seed"])
+    train_subset, valid_subset = random_split(
+        train_dataset, [train_len, val_len], generator=generator
+    )
+    num_workers = hparams["num_workers"]
+
+    if num_workers is None:
+        num_workers = torch.get_num_threads() - 1
+
+    # Create dataloaders
+    train_loader = torch.utils.data.DataLoader(
+        train_subset,
+        batch_size=hparams["batch_size"],
+        shuffle=True,
+        num_workers=num_workers,
+    )
+    valid_loader = torch.utils.data.DataLoader(
+        valid_subset, batch_size=hparams["batch_size"], num_workers=num_workers
+    )
+    test_loader = torch.utils.data.DataLoader(
+        split["test"], batch_size=hparams["batch_size"], num_workers=num_workers
+    )
+
+    return {"train": train_loader, "valid": valid_loader, "test": test_loader}
+
+
+def load_hparams_and_prepare_data(hparams_file, run_opts, overrides):
+    """Load hyperparameters and prepare datasets."""
+    if "SB_YAML_OVERRIDES" in os.environ:
+        overrides = dict(overrides)  # make a copy
+        overrides.update(json.loads(os.environ["SB_YAML_OVERRIDES"]))
+    # Initial hparams load
+    with open(hparams_file) as fin:
+        hparams = load_hyperpyyaml(fin, overrides)
+
+    # Prepare dataset
+    dataset = prepare_dataset(hparams)
+
+    # Update overrides based on actual data shape
+    example_batch = next(iter(dataset))
+
+    overrides.update(
+        T=example_batch["epoch"].shape[1],  # Time dimension
+        C=example_batch["epoch"].shape[0],  # Channel dimension
+        n_train_examples=len(dataset),
+    )
+
+    # Reload hparams with shape information
+    with open(hparams_file) as fin:
+        hparams = load_hyperpyyaml(fin, overrides)
+
```

Comment:
I have found that the conversion of the dataset to a DynamicItemDataset results in a substantial slowdown due to the bandpass filter being applied to each sample on the CPU. One simple remedy  for small datasets could be to do the following:

```suggestion
    # Convert the dataset to a cached dataset
    dataset = InMemoryDataset(dataset)
```

Of course this requires an additional import at the beginning:
```python
from dataio.datasets import InMemoryDataset
```